addons/product/models/product_template.py

```diff
@@ -444,7 +444,12 @@ def _name_search(self, name, args=None, operator='ilike', limit=100, name_get_ui
             args = args if args is not None else []
             products_ns = Product._name_search(name, args+domain, operator=operator, name_get_uid=name_get_uid)
             products = Product.browse([x[0] for x in products_ns])
-            templates |= products.mapped('product_tmpl_id')
+            new_templates = products.mapped('product_tmpl_id')
+            if new_templates & templates:
```

Comment:
This indeed means that we received templates that we explicitly ask to not receive.